<b>Import gurobipy module to perform optimization for linear programming problems</b>

In [36]:
from gurobipy import *

<b>Part 1: American auto marketing cost minimizaiton problem</b>

<b>First, define the Gurobi model</b>

In [37]:
#Gurobi model initialization for American auto problem
auto = Model('auto')

<b>Add Variables to the model and give a name for the variable</b>

In [38]:
#Defining variable for No of Commercials
x = auto.addVar(name="Number of Commercials to run")

In [39]:
#Defining variable for No of Ads
y = auto.addVar(name="Number of Magazine Ads to run")

<b>Set the Objective function</b>

In [40]:
auto.setObjective(500000*x + 750000*y, GRB.MINIMIZE)

<b>Use addConstr method to add constraints to model</b>

In [41]:
#Setting the constraints
auto.addConstr(0.0*x + 0.02*y >= 0.03, name="Sedan_sales_target")
auto.addConstr(0.03*x + 0.01*y >= 0.14, name="SUV_sales_target")
auto.addConstr(-0.01*x + 0.04*y >= 0.04, name="Truck_sales_target")

<gurobi.Constr *Awaiting Model Update*>

<b>Save the model to a file</b>

In [42]:
#Save the model to a file named American-Auto.lp
auto.write('American-Auto.lp')

<b>Perform Optimization to find minimum number of commercials and magazine ads to run to get the target sales</b>

In [43]:
#Use gurobi optimizer to find the optimal solution
auto.optimize()

Gurobi Optimizer version 10.0.1 build v10.0.1rc0 (win64)

CPU model: Intel(R) Core(TM) i5-1035G7 CPU @ 1.20GHz, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 3 rows, 2 columns and 5 nonzeros
Model fingerprint: 0x64581bc0
Coefficient statistics:
  Matrix range     [1e-02, 4e-02]
  Objective range  [5e+05, 8e+05]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e-02, 1e-01]
Presolve removed 1 rows and 0 columns
Presolve time: 0.01s
Presolved: 2 rows, 2 columns, 4 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    3.2083333e+06   1.733333e-01   0.000000e+00      0s
       1    3.5000000e+06   0.000000e+00   0.000000e+00      0s

Solved in 1 iterations and 0.03 seconds (0.00 work units)
Optimal objective  3.500000000e+06


In [44]:
#Get the VAriables
auto.getVars()

[<gurobi.Var Number of Commercials to run (value 4.0)>,
 <gurobi.Var Number of Magazine Ads to run (value 2.0)>]

<b>Finally print the values of variables</b>

In [45]:
for v in auto.getVars():
    print(v.varName, v.x)

Number of Commercials to run 4.0
Number of Magazine Ads to run 2.0


<b>Summary</b>
<p>To achieve the desired sales increase of Sedans, SUV and Trucks<br>
of 3%,14%,4% respectively. American auto needs to run 4 Commercials<br>
and 2 Magazine Ads for a total expense of $3.5 millions</p>

<b>Part 2: Zippy motorcycle marketing cost minimizaiton problem</b>

<b>Model definition</b>

In [55]:
zippy = Model('zippy')

<b>Add Variables to the model</b>

In [56]:
razor = zippy.addVar(name='Number_of_Razor_bike_produced')

In [57]:
zoomer = zippy.addVar(name='Number_of_Zoomer_bike_produced')

<b>Set Objective fucntion to the model</b>

In [49]:
zippy.setObjective(razor + zoomer, GRB.MAXIMIZE)

<b>Add constraints to the model</b>

In [58]:
zippy.addConstr(razor + zoomer <= 700, name='Total no of bikes')
zippy.addConstr(razor - zoomer <= 300, name='Razor bike should not exceeed zoomer')
zippy.addConstr(2*razor + zoomer == 900, name='Total quantity of polymer available')
zippy.addConstr(3*razor + 4*zoomer == 2400, name='Total labor hours available')

<gurobi.Constr *Awaiting Model Update*>

<b>Save the model to a file</b>

In [59]:
zippy.write('Zippy-Bikes.lp')

<b>Use Gurobi optimization</b>

In [60]:
zippy.optimize()

Gurobi Optimizer version 10.0.1 build v10.0.1rc0 (win64)

CPU model: Intel(R) Core(TM) i5-1035G7 CPU @ 1.20GHz, instruction set [SSE2|AVX|AVX2|AVX512]
Thread count: 4 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 4 rows, 2 columns and 8 nonzeros
Model fingerprint: 0xad34d65d
Coefficient statistics:
  Matrix range     [1e+00, 4e+00]
  Objective range  [0e+00, 0e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [3e+02, 2e+03]
Presolve removed 4 rows and 2 columns
Presolve time: 0.01s
Presolve: All rows and columns removed
Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   0.000000e+00   0.000000e+00      0s

Solved in 0 iterations and 0.02 seconds (0.00 work units)
Optimal objective  0.000000000e+00


In [61]:
zippy.getVars()

[<gurobi.Var Number_of_Razor_bike_produced (value 240.0)>,
 <gurobi.Var Number_of_Zoomer_bike_produced (value 420.0)>]

<b>Get the values of variables</b>

In [62]:
for v in zippy.getVars():
    print(v.varName, v.x)

Number_of_Razor_bike_produced 240.0
Number_of_Zoomer_bike_produced 420.0


<b>Summary</b>
<p>Zippy motorcycle can pproduce 240 Razor bikes<br>
and 420 Zoomer bikes for a total profit of $33600</p>